### PR TITLE
Leptos 0.7 upgrade

### PR DESCRIPTION
**Pull request by ogghead**
Hi all!

This will ideally address #44. ~Creating this draft of Leptos 0.7 upgrade with partial changes -- several functions in lib.rs are still not fully updated and so as of time of creation, this does not compile. Creating this in hopes of getting eyes from Leptos experts on whether this is the right path to follow to upgrade. Contributions to move this to completion are appreciated from anyone interested!~

Update 11/4:
leptos_wasi now provides a universal foundation for running a Leptos back-end in Webassembly runtimes -- I have updated this PR with a draft template relying on that integration. Still need to update README with changes, but this is tentatively ready for use

Update 11/10:
Updated README and retested -- everything is working as expected -- ready for feedback!

### CODE REVIEW

**Comment by itowlson** [comment]
@benwis I can't add you as a reviewer for some reason but it would be great if you could also take a look - thanks!

File: README.md
Comment (itowlson) [inline README.md]: If you're going to credit me, you should credit benwis in bigger letters - I started the project but he really made it sing.

File: templates/leptos-ssr/content/Cargo.toml
Comment (itowlson) [inline templates/leptos-ssr/content/Cargo.toml]: Will we be able to relax this bound a bit once Leptos is 0.7 RTM and `leptos_wasi` is on `crates.io`?

File: templates/leptos-ssr/content/src/app.rs
Comment (itowlson) [inline templates/leptos-ssr/content/src/app.rs]: If we wish to change this I would suggest Spin rather than WebAssembly (because Leptos uses WebAssembly on the client anyway).

File: templates/leptos-ssr/content/src/server.rs
Comment (itowlson) [inline templates/leptos-ssr/content/src/server.rs:13]: I wonder if we should use the `#[http_component]` macro (which allows the IncomingRequest/ResponseOutparam signature) instead of the raw binding here?  It may be more familiar to Spin users.  But maybe app developers don't touch this file enough to make any odds.

File: templates/leptos-ssr/content/src/server.rs
Comment (itowlson) [inline templates/leptos-ssr/content/src/server.rs:41]: It's worth commenting that developers need to add new server functions here.  It's quite easy to miss now it's in the middle of a fluent chain.

File: templates/leptos-ssr/content/src/server.rs
Comment (itowlson) [inline templates/leptos-ssr/content/src/server.rs]: I'm a bit wary of using `expect` rather than propagating an error, but maybe there's nothing we can meaningfully do with the error?  Still, I'd prefer to centralise the panic assumption further up - could we have this function return a `Result<()>` and process the falure case in `handle`?

**Comment by itowlson** [comment]
Wow, leptos-wasi makes all this _soooooooooo_ much simpler...!

**Comment by itowlson** [comment]
@ogghead Just had a query on Discord about this.  Would you like this merged at stands with Leptos 7-rc1 (once Ben has looked and any required tweaks/tidying have been done), or were you envisaging waiting for Leptos 7 release?  I am okay with either, just checking on how you feel.

**Comment by ogghead** [comment]
> @ogghead Just had a query on Discord about this.  Would you like this merged at stands with Leptos 7-rc1 (once Ben has looked and any required tweaks/tidying have been done), or were you envisaging waiting for Leptos 7 release?  I am okay with either, just checking on how you feel.
> 
> 

Hey great timing 👋 thanks for reviewing! I think this is _tentatively_ ready to merge barring any additional required tweaks/tidying. I am not sure exactly when the planned release date of Leptos 0.7 is but I would hope any additional changes needed would be minimal. So, merging with these dependencies and creating a quick follow-up PR to update versions + release a tagged version of the template once 0.7 is out makes sense to me!

File: README.md
Comment (ogghead) [reply README.md]: Good callout, oversight on my part! I will update this now!

File: templates/leptos-ssr/content/Cargo.toml
Comment (ogghead) [reply templates/leptos-ssr/content/Cargo.toml]: It has to do with the supported versions of wasi:http interface in Spin -- the latest minor version of wasi uses a version that (when I published this) was unsupported, though it's possible this is outdated now with the release of Spin 3.0

File: templates/leptos-ssr/content/src/app.rs
Comment (ogghead) [reply templates/leptos-ssr/content/src/app.rs]: Ahh, this was some copy pasta from merging the templates. Will update now!

File: templates/leptos-ssr/content/src/server.rs
Comment (ogghead) [reply templates/leptos-ssr/content/src/server.rs:13]: I ran into some challenges with reconciling the types before when I dug into making that work -- I hit a dead-end trying to convert a Spin IncomingRequest into the one expected by leptos_wasi iirc. I will take another crack at this

File: templates/leptos-ssr/content/src/server.rs
Comment (ogghead) [reply templates/leptos-ssr/content/src/server.rs:41]: Good point, I will add this now

File: templates/leptos-ssr/content/src/server.rs
Comment (itowlson) [reply templates/leptos-ssr/content/src/server.rs:13]: @ogghead Oh, if it doesn't work automatically then don't worry about it.  It's no biggie.

File: templates/leptos-ssr/content/Cargo.toml
Comment (itowlson) [reply templates/leptos-ssr/content/Cargo.toml]: I checked in with one of the `wasi` crate maintainers and he believes that the crate (and indeed WASI interface) should be backward compatible with WASI 0.2.0.  If it's not, let me know and I will rope him in.  Thanks!

File: templates/leptos-ssr/content/src/server.rs
Comment (ogghead) [reply templates/leptos-ssr/content/src/server.rs]: Very good point -- handling behavior with Result is more resilient and Rustic. I will update this now

File: templates/leptos-ssr/content/Cargo.toml
Comment (ogghead) [reply templates/leptos-ssr/content/Cargo.toml]: Interesting! Let me give it another shot without this exact dep, I will update shortly

Comment (ogghead) [reply templates/leptos-ssr/content/Cargo.toml]: ~Indeed this appears to run just fine for me now -- not sure what exactly was going wrong before! Updated to use 0.13 without specific version~

When I run with `version = "0.13.3+wasi-0.2.2"`, I see
```
2024-11-19T02:11:57.756986Z ERROR spin_trigger::cli: Trigger executor failed
Error: Expected component to export one of `wasi:http/incoming-handler@0.2.0-rc-2023-10-18`, `wasi:http/incoming-handler@0.2.0-rc-2023-11-10`, `wasi:http/incoming-handler@0.2.0`, `wasi:http/incoming-handler@0.2.1`, or `fermyon:spin/inbound-http` but it exported none of those
```

The handler that _is_ exported is `wasi:http/incoming-handler@0.2.2` in the latest minor version of `wasi`, but perhaps I am missing something

File: templates/leptos-ssr/content/src/server.rs
Comment (ogghead) [reply templates/leptos-ssr/content/src/server.rs:13]: Yeah, still running into challenges around converting types sadly -- I had gone down the route of newtypes to satisfy orphan rule before but it was becoming a rabbithole. This would be ideal though, I can create an issue to track that effort if you're on board with that

File: templates/leptos-ssr/content/Cargo.toml
Comment (itowlson) [reply templates/leptos-ssr/content/Cargo.toml]: Oh!  I wonder if this relates to Spin checking whether it should apply an adapter.  It knows the RCs need adapting, but perhaps it doesn't yet know that 0.2.2 _doesn't_ need adapting.  Don't worry then, leave as is until we can fix in Spin!  Sorry for the bad guidance!

File: templates/leptos-ssr/content/src/server.rs
Comment (itowlson) [reply templates/leptos-ssr/content/src/server.rs:13]: Yeah it is definitely not worth investing more time in.  Again, sorry for sending you in an unsuccessful direction.

**Comment by ogghead** [comment]
Ok, addressed PR feedback thus far, but please let me know if you have any other thoughts @itowlson! I very much appreciate your feedback on this :) @benwis hopefully this PR is up to snuff!

File: templates/leptos-ssr/content/src/server.rs
Comment (ogghead) [reply templates/leptos-ssr/content/src/server.rs:13]: Totally no worries! I spent more time than I would care to admit trying to get that working a couple weeks ago XD I do appreciate the feedback, as I did want to test whether ~Rohan would come to our aid~ Spin 3 or a new version of leptos_wasi might get this rolling

File: templates/leptos-ssr/content/src/server.rs
Comment (itowlson) [inline templates/leptos-ssr/content/src/server.rs]: Should `newleptos` be a template-derived app name here?  cf previous `conf.leptos_options.output_name = "{{project-name | snake_case}}".to_owned();`

**Review by itowlson — APPROVED** [review]
LGTM except for one detail in the template - thanks!

File: templates/leptos-ssr/content/Cargo.toml
Comment (ogghead) [reply templates/leptos-ssr/content/Cargo.toml]: All good! I should have documented this in a comment to help with context -- if you by chance know of a tracking ticket in Spin I can link that on the dependency (and totally no worries either way, I can link the error for now)

File: templates/leptos-ssr/content/src/server.rs
Comment (ogghead) [reply templates/leptos-ssr/content/src/server.rs]: Ahh good catch, I copied some non-templated parts back when I implemented changes, apologies for not reviewing more closely. I will fix now

File: templates/leptos-ssr/content/Cargo.toml
Comment (itowlson) [reply templates/leptos-ssr/content/Cargo.toml]: Tracking at https://github.com/fermyon/spin/issues/2928

Comment (ogghead) [reply templates/leptos-ssr/content/Cargo.toml]: Updated with the link, thank you for making that! I would be interested in helping to address that, though I must admit my knowledge of that facet of Spin is quite limited

Comment (itowlson) [reply templates/leptos-ssr/content/Cargo.toml]: I'm sure one of our contributors who's more knowledgeable than I about that area would be delighted to point you in the right direction and support you learning the ways of it.  Feel free to put your hand up if you have the time and enthusiasm!

**Comment by itowlson** [comment]
@benwis Sorry to badger, but do you know if you'll have bandwidth to look at this?  No worries if not.

File: templates/leptos-ssr/content/src/app.rs
Comment (benwis) [inline templates/leptos-ssr/content/src/app.rs:32]: Is there no css file in this one? It's fine if not, but if so it's removed it. It might be better to keep it

File: README.md
Comment (benwis) [inline README.md:52]: Did you remove this section for a reason?

File: templates/leptos-ssr/content/src/server.rs
Comment (benwis) [inline templates/leptos-ssr/content/src/server.rs]: This is kinda strange, cargo-leptos should be setting this as an env var, and get_configuration(None) should load it you shouldn't be needing to hardcode it

File: README.md
Comment (benwis) [inline README.md]: This whole configuration thing, much like the other one, is somewhat bypassing cargo-leptos's values for dev and the env vars people usually set for prod

**Review by benwis — COMMENTED** [review]
Overall pretty good, left a couple comments

File: templates/leptos-ssr/content/src/server.rs
Comment (itowlson) [reply templates/leptos-ssr/content/src/server.rs]: @benwis This might be historical (it's existing code which GitHub is getting confused and showing as a diff).  During early integration work I found myself needing to force the project name, and I tried a bunch of ways to do that: it's quite likely more of them got left in than are strictly needed...!

File: README.md
Comment (ogghead) [reply README.md]: That's fair -- I think in the context of WASI we may need to configure env vars in a different way, but this could just be a remnant -- I will investigate cleaning these up to match the typical Leptos patterns

File: templates/leptos-ssr/content/src/app.rs
Comment (ogghead) [reply templates/leptos-ssr/content/src/app.rs:32]: I moved this to the [shell function](https://github.com/fermyon/leptos-spin/pull/48/files#diff-782be6a82ee08bd1b4914a1c0ca4dba43937f23e80b5d065f06287678ef51ee1R15) -- I recall that was the pattern I saw in some Leptos examples but I am certainly not a Leptos expert! Let me know if it is preferable to keep the stylesheet in App

File: README.md
Comment (ogghead) [reply README.md:52]: I was seeing this work with expect_context in testing -- for example [in the NotFound component](https://github.com/fermyon/leptos-spin/pull/48/files#diff-782be6a82ee08bd1b4914a1c0ca4dba43937f23e80b5d065f06287678ef51ee1R81)

Perhaps this is still unsupported though and I was misreading that behavior?

File: README.md
Comment (ogghead) [reply README.md]: Indeed a test app seemed to work fine without this override. Pushed up the removal here + in codegen, good callout!

File: templates/leptos-ssr/content/src/server.rs
Comment (ogghead) [reply templates/leptos-ssr/content/src/server.rs]: Good callout, a test Leptos Spin app does seem to function fine without this -- I went ahead and removed here and in README.

Thanks for the context @itowlson -- do you remember why the project name had to be forced?

**Comment by ogghead** [comment]
Thanks for the review @benwis and the bump @itowlson! I have responded to comments + pushed up changes to address. I will keep an eye out for additional feedback + discussion on open comments -- I appreciate the eyes on this PR!

File: templates/leptos-ssr/content/src/server.rs
Comment (itowlson) [reply templates/leptos-ssr/content/src/server.rs]: @ogghead I think it was to do with finding assets under the `/pkg` route - it was looking for assets using the wrong name.  But yeah this was very much an "at the time" thing.  I am happy to write it off as an accident of history that is no longer needed.  Thanks for testing and thanks @benwis for querying it!

File: README.md
Comment (benwis) [reply README.md:52]: We definitely still do route generation that could not have it, but depending on what you're getting it may or may not work. Request or something from Request is the most likely to hit this issue. Assuming it still is, this could be reworded and clarified

File: templates/leptos-ssr/content/src/app.rs
Comment (benwis) [reply templates/leptos-ssr/content/src/app.rs:32]: That's a good way to do it, I missed that

**Comment by benwis** [comment]
> Thanks for the review @benwis and the bump @itowlson! I have responded to comments + pushed up changes to address. I will keep an eye out for additional feedback + discussion on open comments -- I appreciate the eyes on this PR!

Really nice work! Only one comment I see left, otherwise this should be good to merge

File: README.md
Comment (ogghead) [reply README.md:52]: Ah good to know! Updated to re-add this section with a bit of additional context

**Comment by ogghead** [comment]
I appreciate the quick review and additional context @benwis! Updated to address that comment now

**Review by benwis — APPROVED** [review]
LGTM

**Comment by itowlson** [comment]
Thanks @ogghead for your work on this and @benwis for bringing your knowledge and attention to detail!  Hopefully future updates will be easier now that Leptos has a WASI integration within the org!

**Pull request merged**
